experimental/rule_inference/rule_application.py

```diff
@@ -0,0 +1,70 @@
+import attr
+import toml
+from typing import List
+from polyglot_piranha import (
+    Rule,
+    PiranhaArguments,
+    RuleGraph,
+    execute_piranha,
+    PiranhaOutputSummary,
+)
+
+
+@attr.s
+class CodebaseRefactorer:
+    """
+    A class that uses Piranha to refactor an entire codebase based on rules specified in a .toml file.
+    """
+
+    language = attr.ib(type=str)
+    path_to_codebase = attr.ib(type=str)
+    path_to_rules = attr.ib(type=str)
+    include_paths = attr.ib(type=List[str], default=None)
+    exclude_paths = attr.ib(type=List[str], default=None)
+
+    def refactor_codebase(self, dry_run: bool = True) -> List[PiranhaOutputSummary]:
+        """
+        Applies the refactoring rules to the codebase.
+        Returns a list of piranha summaries
+        """
+        # Load the rules from the .toml file
+        with open(self.path_to_rules, "r") as file:
+            toml_dict = toml.load(file)
+
+        rules = toml_dict.get("rules", [])
+        if not rules:
+            raise Exception("TOML does not include any rule specifications.")
+
+        # Create the Piranha rule graph
+        rule_graph = self.create_rule_graph(rules)
+
+        # Create the PiranhaArguments object
+        args = PiranhaArguments(
+            language=self.language,
+            path_to_codebase=self.path_to_codebase,
+            rule_graph=rule_graph,
+            dry_run=dry_run,
+        )
+
+        # Execute the refactoring
+        return execute_piranha(args)
+
+    @staticmethod
+    def create_rule_graph(toml_rules: List[dict]) -> RuleGraph:
```

Comment:
so we dont use raw toml files, we create a rule graph in python API ? 
(makes sense to me, just confirming the choices)

Reply:
Yes. the toml is just an intermediate representation for the model. I think it;s easier to express the rule in toml (avoid python boilerplate and so on)